DHFPROD-4625: Hub Central Security functional spec

ryanjdew opened the pull request:
### Description

#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+ - data-hub-load-data-reader
+ - data-hub-mapping-reader
+ - data-hub-match-merge-reader
+ - data-hub-step-definition-reader
```
IIRC, a hub-central-step-runner should only be able to see the steps in a flow(Run tile) but not the step configurations(Entities tile). So should not inherit load-data/mapping/match-merge/step-definition reader.

ryanjdew replied:
That is a difference between the `hub-central-`prefixed roles and the `data-hub-` prefixed roles. The `hub-central` roles dictate what can be done through the UI/middle-tier and the `data-hub` roles are primarily document-level permissions. In order for a `hub-central-step-runner` to be able to run a step they need the document-level read permissions to the various artifacts.

rjrudin commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+tags: [security]
+---
+
+# Data Hub Security
```
This is great information, but I'm wondering if it should go into github yet. I don't believe we have any docs like this in github; currently, I think this information should go into a design spec on our internal Wiki. That's primarily for consistency. I do think there's long-term value in exposing this value to the public, I'm just not at the moment how best to do that - i.e. what's the right mix of info in github vs in our public docs? So for now, I think it's best to put this into a "Hub Central Security" design spec page on the Wiki.

rjrudin commented on specs/reference/data-services/security.v1.json:
```diff
@@ -0,0 +1,56 @@
+{
+  "openapi": "3.0.0",
+  "tags":[{
+    "name": "security"
```
I know we have a "SecurityService" already (I may have been the one to suggest that name, can't remember), but I'm wondering if we should rename it to "UserService" because "getAuthorities" is specific to a user. I'm wondering what other functions we would add to this... and I'm not sure. So I don't really know a good name, but I lean towards UserService since this is currently specific to the user calling it. Thoughts?

ryanjdew replied:
Makes sense. Leaves the door open for Hub Central to do some limited security management, etc. in the future.

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+#### Inherited Roles
+ - hub-central-load-reader
+ - data-hub-common-writer
+ - data-hub-load-data-reader
```
hub-central-load-reader inherits data-hub-load-data-reader already.

ryanjdew replied:
good catch. that should be `data-hub-load-data-writer`

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+#### Inherited Roles
+ - hub-central-mapping-reader
+ - data-hub-common-writer
+ - data-hub-mapping-writer
```
hub-central-mapping-reader inherits data-hub-mapping-writer

ryanjdew replied:
Why would hub-central-mapping-reader inherit data-hub-mapping-writer?

bsrikan replied:
My bad. I read data-hub-mapping-writer as reader

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+#### Inherited Roles
+ - hub-central-match-merge-reader
+ - data-hub-common-writer
+ - data-hub-match-merge-writer
```
hub-central-match-merge-reader inherits data-hub-match-merge-writer

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+ - hub-central-mapping-reader
+ - data-hub-common-writer
+ - data-hub-step-definition-writer
+ - data-hub-modules-writer
```
Should hub-central-custom-writer inherit hub-central-custom-reader, so we can do away with inheriting data-hub-step-definition-writer. 
Typo in modules-writer, should be data-hub-module-writer. Should data-hub-common-writer inherit data-hub-module-writer?

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+ - hub-central-custom-writer
+ - hub-central-downloader
+
+### hub-central-custom-writer
```
Typo, should be hub-central-operator

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+
+Can create any type of step and run steps through Hub Central
+
+#### Inherited Roles
```
Should a HC developer inherit HC operator like its data-hub-* counterpart?

ryanjdew replied:
yes

bsrikan commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,266 @@
+
+### hub-central-clear-data
+
+Role to determine if a user can clear data through Hub Central.
```
clear data from Staging, Final and Jobs only? Will this role be allowed to delete provenance?
Are these the artifacts this role is allowed to delete: /entities,/flows(user and default?), /step-definitions(user and default?), /loadData, /mappings, /matching, activity-settings, /jobs, /batch, all the ingested/mapped/mastered documents including triples/graph.

rjrudin replied:
So I said "clear data" yesterday, and now I'm regretting it. :) 

The problem with "clear data" is that I think the expectation from clearing a database is that it's empty. But that won't be the case - the DH and user artifacts will still be present. We expect to do a clear first, and then add those artifacts back, but the net result is not a "clear". 

I think "clear user data" would work. I'd prefer if we had a noun here for consistency with other roles, but I don't have a good suggestion. Think we'll have to accept some verbs here, which is consistent with ML roles - many are nouns, some are verbs.

rjrudin replied:
Also, I think "clear user data" to me implies that it applies to staging/final/jobs.

bsrikan commented on marklogic-data-hub/src/main/resources/hub-internal-config/security/roles/data-hub-common-writer.json:
```diff
@@ -0,0 +1,30 @@
+  "description": "Contains a common set of privileges/roles needed for Data Hub 5+ write operations",
+  "role": [
+    "data-hub-common",
+    "rest-writer"
```
do we need to inherit data-hub-module-writer here?

ryanjdew replied:
no. that isn't a common write permission that we want to grant to any user what has write privileges

rjrudin commented on specs/docs/security/DataHubSecurity.md:
````diff
@@ -0,0 +1,267 @@
+
+#### Example xdmp.securityAssert call
+```javascript
+xdmp.securityAssert("http://marklogic.com/data-hub/function-access", "execute");
````
I'm all for guarding every DS endpoint with this. On one project, we wrote a little tool to parse the xdmp.securityAssert call on the first line that it was found in an endpoint module and then constructed an HTML table output of every endpoint + the privilege that it required. Very handy, may be worth doing in the future.

rjrudin commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,267 @@
+Data Hub role containing the common privileges across Data Hub roles.
+
+#### Inherited Roles
+ - data-hub-module-reader
```
I think we need rest-extension-user here as well. I ran into some surprising issues that were new to me, where if any module imported by a REST extension doesn't have at least one common read permission with every other module, an error will be thrown. So we guarantee that rest-extension-user is always present, and thus I think data-hub-common should have it.

ryanjdew replied:
added the rest-extension-user role here

rjrudin commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,267 @@
+
+#### Inherited Roles
+ - data-hub-common
+ - rest-writer
```
We don't want to inherit this role, because it includes default permissions that work against document-level security. Need to inherit the rest-writer privilege instead.

ryanjdew replied:
Changed to use the privilege.

rjrudin commented on specs/docs/security/DataHubSecurity.md:
```diff
@@ -0,0 +1,267 @@
+Role to determine if a user can download the Hub Central project.
+
+#### Associate Authorities
+ - canDownloadProject
```
I think "canDownloadUserArtifacts" may be better here, since the entire project cannot be downloaded, just what we internally call "user artifacts".

rjrudin replied:
Also, I don't know yet if this role should have all the read-access roles for user artifacts, or if we'd used an amp in the DS endpoint to grab all of them. The latter is certainly possible - I did a little prototype of using xdmp.zip. So a DS endpoint that requires a certain privilege and then uses an amp to construct the zip is possible. But I don't know if that's worth calling out here - it would seem a little early to do that.

ryanjdew commented:
@rjrudin Addressed the issues you've mentioned and I've created the following internal functional spec wiki: https://wiki.marklogic.com/display/ENGINEERING/Hub+Central+Security+Design+Specification. It naturally does quite a bit of referencing to other places we've documented things. 

I think it is important to have information about our security approach as part of the repo, rather than an internal wiki so there is a reference for potential contributors.
